File: backend/app/schemas/ws.py
"""
WebSocket 推送消息 Schema
"""
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime


# === WebSocket 消息类型 ===
class WSMessageType:
    """WebSocket 消息类型常量"""
    ROBOT_STATUS = "robot_status"
    ROBOT_POSITION = "robot_position"
    ROBOT_TORQUE = "robot_torque"
    ALARM_UPDATE = "alarm_update"
    CONNECTION_UPDATE = "connection_update"
    SAFETY_ALERT = "safety_alert"
    ERROR = "error"
    PACKET_LOG = "packet_log"


class WSMessage(BaseModel):
    """WebSocket 通用推送消息结构"""
    type: str
    timestamp: str  # ISO 8601 UTC
    data: Dict[str, Any] = Field(default_factory=dict)


# === 各类型具体数据 ===

class WSRobotStatusData(BaseModel):
    """机器人核心状态"""
    servo_on: bool = False
    running_mode: str = "unknown"
    alarm_active: bool = False
    speed_percent: int = 0
    executing_program: Optional[str] = None


class WSRobotPositionData(BaseModel):
    """机器人位置数据"""
    joints: Dict[str, float] = Field(default_factory=dict)
    end_coords: Dict[str, float] = Field(default_factory=dict)


class WSRobotTorqueData(BaseModel):
    """各轴力矩数据"""
    torques: List[float] = Field(default_factory=list)


class WSAlarmData(BaseModel):
    """报警数据"""
    alarms: List[Dict[str, Any]] = Field(default_factory=list)
    has_active_alarm: bool = False


class WSConnectionData(BaseModel):
    """连接状态数据"""
    connected: bool = False
    last_heartbeat: Optional[str] = None


class WSSafetyAlertData(BaseModel):
    """安全告警数据"""
    alert_type: str
    message: str
    severity: str = "warning"  # critical / warning / info

File: frontend/src/types/ws.ts
// ===== WebSocket 消息类型定义 =====

/** WS 消息类型 */
const WSMessageType = {
  ROBOT_STATUS: 'robot_status',
  ROBOT_POSITION: 'robot_position',
  ROBOT_TORQUE: 'robot_torque',
  ALARM_UPDATE: 'alarm_update',
  CONNECTION_UPDATE: 'connection_update',
  SAFETY_ALERT: 'safety_alert',
  ERROR: 'error',
  PACKET_LOG: 'packet_log',
} as const

type WSMessageTypeValue = typeof WSMessageType[keyof typeof WSMessageType]

/** WS 推送消息 */
interface WSMessage<T = Record<string, unknown>> {
  type: WSMessageTypeValue
  timestamp: string
  data: T
}

/** 机器人状态数据 */
interface WSRobotStatusData {
  servoOn: boolean
  runningMode: string
  alarmActive: boolean
  speedPercent: number
}

/** 位置数据 */
interface WSRobotPositionData {
  joints: Record<string, number>
  endCoords: Record<string, number>
}

/** 力矩数据 */
interface WSRobotTorqueData {
  torques: number[]
}

/** 报警数据 */
interface WSAlarmData {
  alarms: unknown[]
  hasActiveAlarm: boolean
}

/** 连接数据 */
interface WSConnectionData {
  connected: boolean
  lastHeartbeat: string
}

/** 安全告警 */
interface WSSafetyAlertData {
  alertType: string
  message: string
  severity: 'critical' | 'warning' | 'info'
}

export { WSMessageType }
export type {
  WSMessageTypeValue, WSMessage,
  WSRobotStatusData, WSRobotPositionData, WSRobotTorqueData,
  WSAlarmData, WSConnectionData, WSSafetyAlertData,
}
